### PR TITLE
Themes: Add Dialog when Switching From Retired Theme

### LINKED ISSUE #40310: Themes: add a friendly warning when switching away from a retired theme
Suggested by @jjchrisdiehl (on this internal thread: pbg9X-epe-p2)

retired_theme

The task is to adapt the current "Theme activation notice" to display a friendly message to site owners just before they take the destructive action to switch away from a retired theme.

The action isn't reversible for "Unlimited Theme Usage" on Premium, Business, and eCommerce plans. (Site owners who purchase a single theme can get it back in WP Admin theme viewer.)

@senff Adds this note:

> Avoid this from happening in the first place by making it clear that switching away from a retired theme is a permanent action and can not be undone... it might help those who accidentally switch.

#### Suggested wording

Your active theme **Lodestar** is retired. If you activate a new theme, you might not be able to switch back to Lodestar.

### PR DESCRIPTION
**Pull request by Aurorum**
#### Changes proposed in this Pull Request

This converts the dialog that displays when activating a theme that has a Homepage Layout to one that serves that purpose, as well as warning when switching from a retired theme. I don't think that I'm able to change a few other cases of the previous name as it appears they involve conditions outside this repo, but I don't think that's a big deal.

#### Testing instructions

I took immense care here to ensure that nothing is changed when switching to a Homepage Layout, but feel free to verify this just in case! The same dialog should still appear.

This just adds a message when switching to a retired theme. Select a retired theme through WP-Admin (eg. _Harmonic_) and attempt to activate another theme. This should appear, and clicking the primary button should activate the theme as usual. Verify that this dialog doesn't appear in other cases.

Screenshot 2020-04-05 at 18 49 01

cc @lancewillett 

Fixes #40310

### CODE REVIEW

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (a8ci18n) [inline client/my-sites/themes/theme-activation-confirmation-modal.jsx:126]: :ok: This change will be queued for retranslation. We'll use the existing translations in the meantime.

**Comment by Aurorum** [comment]
Bit stick on fixing the remaining three tests; it isn't the only thing dispatched, but I'm not sure how to ensure the tests recognise that.

https://github.com/Automattic/wp-calypso/blob/acdd5ab6d6d8c2635769b3dacfcf374f7de45013/client/state/themes/test/actions.js#L651

**Comment by obenland** [comment]
@Aurorum Would you be up for a refresh on this one?

**Comment by Aurorum** [comment]
Will try to give it a go! There's a lot of conflicts though, so I think this will take some time (or it might just be worth creating a new branch instead).

**Comment by sixhours** [comment]
👋 Hey @Aurorum ! Let us know if you want to try rebasing this or just create a new PR. :)

**Comment by Aurorum** [comment]
Yes, sorry, I'll probably fix this up over the weekend (assuming it's not urgent!) :)

**Comment by kwight** [comment]
@Aurorum No, certainly not urgent, we're happy for whatever time you have to contribute. ❤️

**Comment by Aurorum** [comment]
Apologies to Manage for the ping - this is unrelated to My Home, but I accidentally used that branch to update this PR without realising it'd ping people.

**Comment by Aurorum** [comment]
This should be working again. :) 

Screenshot 2021-05-22 at 14 54 15

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (sixhours) [inline client/my-sites/themes/theme-activation-confirmation-modal.jsx]: Changing this event name will disconnect it from data we've collected since #52926 was deployed (just last week, so not much).

But this does make me wonder if we should have a more generic event name and use props to track the type of modal we're viewing (`retired` theme warning, `auto-loading homepage`, etc.) /cc @kwight @ianstewart any preferences here?

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (sixhours) [inline client/my-sites/themes/theme-activation-confirmation-modal.jsx:135]: This event name hasn't been changed to match the new format.

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (sixhours) [inline client/my-sites/themes/theme-activation-confirmation-modal.jsx]: It looks like we have a selector for this: https://github.com/Automattic/wp-calypso/blob/trunk/client/state/themes/selectors/get-active-theme.js

And I think we already have the data for `installingThemeName` as part of `theme`, rather than using `get`.

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (sixhours) [inline client/my-sites/themes/theme-activation-confirmation-modal.jsx]: I think you can simplify the args like you have with the others:

```args: { themeName }```

**Review by sixhours — CHANGES_REQUESTED** [review]
Thanks @Aurorum I took this for a spin! Looks like a good start, I left some code comments that are relatively minor. Visually, I noticed the buttons are aligned to the left rather than centered in the modal:

Screen Shot 2021-05-25 at 2 53 18 PM

Functionally, when you switch from a retired theme to an auto-loading homepage theme (ie. one of the Recommended themes), you don't get the option to use your own content or replace the homepage with the demo site's. I think we need to keep that feature regardless of whether you're switching from a retired theme or not. I'm not sure what that flow looks like, though -- a second modal prompt after the retirement notice? Or maybe it's not a big deal... /cc @ianstewart in case you have thoughts (I know you'll have thoughts :D).

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (kwight) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: I don't think we need to go there – we can always use unique names (even if they get a little longer).

Comment (sixhours) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: OK, so it sounds like we should update this and record a separate Tracks event for the retired theme modal rather than changing the autoloading homepage event.

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (sixhours) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx:135]: Instead of updating this event name, let's record a new Tracks event when the retired warning modal is dismissed so it can be tracked separately.

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: `getActiveTheme` returns the ID rather than the name (ie. `twentytwentyone` instead of `Twenty Twenty-One`), but I think that you're right about there being a much better way to do this using `getCanonicalTheme` - done so now! :)

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: Not in this case, I'm afraid. If the theme is retired, `themeName` is the name of the active theme. So in reality, you'd be asking a user if they want to activate their current theme.

That being said, I completely agree that this is a little confusing so I've switched `theme` to `installingTheme`.

**Comment by Aurorum** [comment]
Thanks @sixhours! All the issues (except trackEvents) should be resolved now. I've come up with this for cases when the current theme is retired and the selected theme should display the homepage options.

Screenshot 2021-05-26 at 18 44 07

Let me know what you think. :)

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: Sure thing, but assuming that this design is okay, which one should be fired when the modal in https://github.com/Automattic/wp-calypso/pull/40793#issuecomment-848902352 is displayed?

Comment (sixhours) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: I think we can leave all the `autoloading_homepage` Tracks events as-is, and add a new event for the retired theme warning (a `*_view` event when the retired theme warning message is shown on the autoloading homepage modal). Then when looking at Tracks flows we can segment by folks who've seen both the autoloading homepage modal AND the retired theme warning at the same time.

I don't think it's necessary to add separate `click` and `dismiss` events to the buttons when the warning is displayed on the autoloading homepage modal, though.

Let me know if that makes sense. :D

File: client/my-sites/themes/theme-activation-confirmation-modal.scss
Comment (sixhours) [inline client/my-sites/themes/theme-activation-confirmation-modal.scss]: Let's replace this with the new Gutenberg/WordPress core breakpoints system: https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/css.md#media-queries

File: client/my-sites/themes/theme-activation-confirmation-modal.scss
Comment (sixhours) [inline client/my-sites/themes/theme-activation-confirmation-modal.scss]: Can this use `font-size: $font-body-small` instead?

**Review by sixhours — COMMENTED** [review]
This is working well for me! Left a couple small notes on the CSS, and the Tracks events need to be finalized and tested.

File: client/my-sites/themes/theme-activation-confirmation-modal.scss
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.scss]: I don't think these styles are even used - they're from the old stylesheet, but there's no corresponding element with a class name of `*-hint`, so I've just removed them.

File: client/my-sites/themes/theme-activation-confirmation-modal.scss
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.scss]: Done! :)

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: I think that I've implemented those changes now - let me know if I've misunderstood! :)

**Comment by kwight** [comment]
> "You might not be able..." 

@sixhours Should this be "You will not be able..." – is there any situation in which someone could re-activate their retired theme?

**Comment by Aurorum** [comment]
My understanding is that the usual scenario is people can re-activate their retired theme. For example, you can activate a retired theme now by going to `/customize/site?theme=pub/harmonic`.

update: oops, sorry, just saw this wasn't directed at me!

**Comment by sixhours** [comment]
No worries @Aurorum, you're right!

> @sixhours Should this be "You will not be able..." – is there any situation in which someone could re-activate their retired theme?

Yes, retired free themes can be re-activated from `/wp-admin`, but retired premium themes cannot.

**Comment by ianstewart** [comment]
@sixhours @Aurorum I think my only note was that the original issue had a link to documentation in the modal that offered more support/explanation for retiring a theme. It seemed useful.

**Comment by Aurorum** [comment]
I don't think that such documentation relating to retired themes actually exists unless I've missed it - please let me know the link if there's one which I should include!

**Comment by ianstewart** [comment]
This was the mockup from [the issue I was thinking of](https://github.com/Automattic/wp-calypso/issues/40310):

image

Link referenced here: 

https://github.com/Automattic/wp-calypso/issues/40310#issuecomment-832262506

Looks like @kathrynwp did a cleanup on it here:

https://github.com/Automattic/wp-calypso/issues/40310#issuecomment-843457240

Thanks for taking a look @Aurorum!

**Comment by Aurorum** [comment]
Ah - thank you, I didn't realise a section on Retired Themes had been added! I've included it in the dialog now.

**With homepage options:**
Screenshot 2021-06-03 at 07 56 47

**Without homepage options:**
Screenshot 2021-06-03 at 07 59 40

**Comment by mreishus** [comment]
I believe the `isVisible` prop is no longer being used; which means the `wasVisible` state and `getDerivedStateFromProps()` function aren't doing anything and can be safely removed.
This was causing the dialog selection to be reset when closing and re-opening the dialog, which it no longer does (I don't think this is a big deal). Only pointing out that for simplicity we should either remove the dead code or re-implement the feature. I think removing the dead code is OK.

**Comment by sixhours** [comment]
Phew, this is a doozy. :D This is my attempt to get it all straight in my head, with a screenshot of the dialog, and the Tracks events when you:

* View the modal (if there is one)
* Click "Activate"
* Click "Keep my current theme"

________

### Retired -> Autoloading theme

Screen Shot 2021-06-03 at 10 57 52 AM

View

retiredandautoloading

Activate

Screen Shot 2021-06-03 at 11 05 56 AM

**_(Note: This doesn't distinguish between retired or non-retired themes; is that intended?)_**

Keep current theme/dismiss

Screen Shot 2021-06-03 at 11 06 42 AM

**_(Note: This doesn't distinguish between retired or non-retired themes; is that intended?)_**

________

### Retired -> Non-autoloading theme

Screen Shot 2021-06-03 at 10 58 06 AM

View 

retiredandnonautoloading

**_(Note, this Tracks event name doesn't make a distinction between retired or non-retired themes. Is that intended?)_**

Activate

Screen Shot 2021-06-03 at 11 02 09 AM

Keep current theme

Screen Shot 2021-06-03 at 11 07 24 AM

________

### Non-retired -> Autoloading theme

Screen Shot 2021-06-03 at 10 57 27 AM

View

autoloading

**_(Note: This event name no longer matches the one in production -- `calypso_theme_autoloading_homepage_modal_view` -- and I believe it should.)_**

Activate

Screen Shot 2021-06-03 at 11 08 46 AM

Keep current theme

Screen Shot 2021-06-03 at 11 08 20 AM

________

### Non-retired -> Non-autoloading theme

[No warning shown as expected.]

________

I think this is close, but the Tracks event names aren't quite consistent. That said, it doesn't matter what they are as long as they're useful to the folks building funnels :) @ianstewart @kwight maybe you can take a look and confirm. Anything I've noted above in **_bold italics_** is a question.

**Comment by Aurorum** [comment]
@mreishus Nice spot, fixed - thanks! :)

**Comment by Aurorum** [comment]
@sixhours 

> (Note: This event name no longer matches the one in production -- calypso_theme_autoloading_homepage_modal_view -- and I believe it should.)

Fixed!

> (Note, this Tracks event name doesn't make a distinction between retired or non-retired themes. Is that intended?)

My understanding is that this is what you advised in https://github.com/Automattic/wp-calypso/pull/40793#discussion_r641723371 - please let me know if I was mistaken!

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (kwight) [inline client/my-sites/themes/theme-activation-confirmation-modal.jsx]: Hm, this will be problematic. We don't want to rename events that are still valid, otherwise we lose all history and usefulness of that event. We also don't want to use string interpolation in event names; it makes them very difficult to find in the codebase.

We can record multiple Tracks events at the same time – can we leave the existing ones as they are, then add another `recordTracksEvent` call specifically for the retired stuff? It's fine that the button language or presentation may have needed to change, as long as the action taken matches with the event fired.

**Comment by ianstewart** [comment]
> That said, it doesn't matter what they are as long as they're useful to the folks building funnels

Good reminder that should make sure we document tracks events as we work with them. Especially when the naming might be off. :)

File: client/my-sites/themes/theme-activation-confirmation-modal.jsx
Comment (Aurorum) [reply client/my-sites/themes/theme-activation-confirmation-modal.jsx]: Sure thing, let me know if the new ones are okay!

**Comment by kwight** [comment]
@Aurorum My apologies, we're going to have to put a hold on this for the moment – the original homepage content modal is now under a re-design, so we should let that dust settle before re-evaluating how best to implement this (should be in the next few weeks). Sorry again, thanks – we'll revisit this soon!

**Comment by kwight** [comment]
@ianstewart @onuro @blackjackkent Now that we have the theme switch preview modal merged, how do you all think we should handle this case where we want to inform the user of retired theme considerations too? The modal could be getting kinda overloaded, but having multiple modals doesn't feel like a great experience either. 🤔

**Comment by onuro** [comment]
I would suggest working on the wording rather than the design here. Word "retired" could mean a vague expression to users.

How about "Your current theme: **Libre** is an old theme. When you switch to {theme name}(modern theme), you might not be able to switch back to Libre"?.

Final phrase is open for discussion, but my take on this is avoiding intimidation as much as possible.

**Comment by kwight** [comment]
@onuro But one modal to handle both cases sounds like the way to go? Hm, yeah, "old" can seem like we're selling ourselves and the user choice short. 🤔 @kathrynwp @jordesign @MindyPostoff Any ideas to suggest?

**Comment by kathrynwp** [comment]
To avoid confusion, I would strongly suggest we stick with the word we've always used with users ("retired"), instead of coming up with something new. 

And we could also link from the warning to the section about retired themes, as @ianstewart suggested earlier:
https://wordpress.com/support/changing-themes/#retired-themes

**Comment by MindyPostoff** [comment]
> Any ideas to suggest?

This is tricky. Do we want people to switch away from a retired theme? If so, I think our modal should excite the user to change themes by highlighting the positives, while also explaining the cost of doing so. Something like:

> [New Theme Name] is great choice to help your site look more modern! Please note that [Retired Theme Name] is retired and no longer supported, so you might not be able to switch back.

Or something along those lines?

**Comment by kwight** [comment]
@Aurorum 👋 Now that our latest round of that modal is live, would you like to take another crack at updating this? So we'll stick with the same approach you had, of the existing modal integrating this message.

**Comment by Aurorum** [comment]
I've updated this PR with a new set of changes that should work with the updated modal. I haven't really explored the new theme modal in detail before now though so please let me know if there's something wrong!

@kwight @MindyPostoff @onuro @kathrynwp Could you please advise if the wording in https://github.com/Automattic/wp-calypso/pull/40793#issuecomment-876498932 is final? I'm happy to switch it to that if so! 

By the way, is there an issue with having two "Learn more" links by each other? I feel it might be more clean to have the retired message above the images in this scenario, but I'll leave it up to someone else to decide. :) 

Screenshot 2021-07-09 at 13 53 33

**Comment by kwight** [comment]
> By the way, is there an issue with having two "Learn more" links by each other? I feel it might be more clean to have the retired message above the images in this scenario, but I'll leave it up to someone else to decide. :)

Whatever @onuro thinks is best. 👍 

> Could you please advise if the wording in #40793 (comment) is final?

Hm, we're getting pretty close to too much text again. 🤔  Let's go with the second part of @MindyPostoff 's comment: "Please note that [retired theme name] is retired and no longer supported, so you might not be able to switch back."

Thanks @Aurorum !

**Comment by onuro** [comment]
I'm in favor of @Aurorum's top placement(after the heading) of the message within [maybe a card](https://wpcalypso.wordpress.com/devdocs/design/card)/or plain.

image

**Comment by Aurorum** [comment]
Thank you! I quite like the card, and I've included it in the latest commits along with the text changes. :) 

Screenshot 2021-07-13 at 19 50 42

File: client/my-sites/themes/install-theme-button.scss
Comment (sixhours) [inline client/my-sites/themes/install-theme-button.scss:2]: Hmm, I feel like the linter should have caught this. Looks like it's reverted from using tab indentation to spaces.

**Review by sixhours — CHANGES_REQUESTED** [review]
Screen Shot 2021-07-14 at 11 31 59 AM
Screen Shot 2021-07-14 at 11 37 10 AM

This is mostly working for me! I appreciated that you took the time to rename the components/selectors/etc. to something that's more understandable. :D A few gotchas:

* I double-checked all combinations of the Tracks events (retired/not retired/autoloading/not autoloading). I noticed the Tracks events for a theme that only shows the retired message correspond are still the "autoloading_homepage_" events. Not sure if that case should be counted separately? /cc @themeshaper  since I'm not sure what format for the event names/properties makes the most sense to you.
* This probably needs to be rebased since we've been working on the Showcase a bunch lately; I tried to do so locally and got conflicts.
* I ran into a bug with specific themes on the Recommended tab:

Screen Shot 2021-07-14 at 11 35 24 AM

Clicking on "Activate" doesn't bring up a modal or activate the theme. Those themes work fine on the All Themes tab, and on production. Not sure what's going on there. There are no errors in the console.

**Comment by sixhours** [comment]
> I ran into a bug with specific themes on the Recommended tab:

With more investigation I think this might be related to another bug in production. Those themes can still be activated on production, but they don't bring up the theme switch modal as expected. Opening an issue for it in #54559

File: client/my-sites/themes/install-theme-button.scss
Comment (Aurorum) [reply client/my-sites/themes/install-theme-button.scss:2]: I don't remember the linter running here, but it seems to have corrected to tab indentation? At least, I can't seem to revert it.

Screenshot 2021-07-15 at 21 27 41

**Comment by Aurorum** [comment]
Thanks @sixhours! :) 

> I double-checked all combinations of the Tracks events (retired/not retired/autoloading/not autoloading). I noticed the Tracks events for a theme that only shows the retired message correspond are still the "autoloading_homepage_" events. Not sure if that case should be counted separately? 

I've tried to make the Tracks events reflect what's outlined in https://github.com/Automattic/wp-calypso/pull/40793#discussion_r641723371, but please let me know if I've misunderstood!

> This probably needs to be rebased since we've been working on the Showcase a bunch lately; I tried to do so locally and got conflicts.

Done!

> I ran into a bug with specific themes on the Recommended tab:

I did try to look into this, but I can't really see the cause of it. Given that it's something ongoing in `trunk` already, is it in the scope of this PR though?

**Comment by sixhours** [comment]
Thanks Aurorum! We really want to get this pushed out before it sits too long. 

> I did try to look into this, but I can't really see the cause of it. Given that it's something ongoing in trunk already, is it in the scope of this PR though?

Not at all, but I'm concerned that theme activation stops working entirely for these themes with this PR. :-/ So I think this is blocked until we can address #54559 .  I'll take a look at it today and see if I can figure it out!

File: client/my-sites/themes/install-theme-button.scss
Comment (sixhours) [reply client/my-sites/themes/install-theme-button.scss:2]: 🤷‍♀️ Obvs not a blocker. Ignoring. :D

**Review by sixhours — APPROVED** [review]
The activation glitch is resolved when I apply the fix from #54648 and this is still working beautifully, so I'm going to run the e2e tests and ship it provided everything passes. Great work @Aurorum , thanks for sticking with us on this one. :)

**Comment by sixhours** [comment]
This may need a rebase again. 🙃 I'm getting a build error and some test failures. Let me know if you need additional info @Aurorum 
Screen Shot 2021-07-16 at 10 56 54 AM

**Comment by Aurorum** [comment]
Thanks @sixhours! Uhm, I can't immediately see a solution to that. Is there any way which I'm able to run those tests locally? I've tried updating the branch and rewriting the `isUsingRetiredTheme` selector slightly, but I don't really understand the problem that the tests are designed to warn against.

**Comment by mreishus** [comment]
@Aurorum : Try `yarn test-client:watch` in one terminal, then changing a comment in `client/state/themes/test/actions.js` to trigger those tests to run. That worked for me. `yarn test-client` should also work but it's pretty slow to run all of them.

**Comment by Aurorum** [comment]
Thanks @mreishus! 

I've been looking into this a lot over the weekend, and I don't feel like I'm any closer to identifying the issue; even if you remove the `getActiveTheme` usage in `isUsingRetiredTheme`, an error still appears. There aren't any explicit references to the auto loading homepage modal or anything being changed here, so I don't understand what exactly is causing the tests to fail here - would anyone have any ideas by any chance? :)

**Comment by sixhours** [comment]
I've been looking into this, and my best guess is it's some conflict with using `getActiveTheme` within the theme activation action. Maybe due to the lag between the API and the state documented here:

Screen Shot 2021-07-21 at 4 27 30 PM

That said, I haven't been able to figure it out for sure. 🙃

**Comment by mreishus** [comment]
client/state/themes/actions/activate.js -> Change line 44 to `themeHasAutoLoadingHomepage( getState(), themeId ) &&` -> Test starts passing.

I'm not completely sure what's going on, but it's possibly that calling activate on a theme in a test used to fetch its data, but after the changes made, it does not, it only shows a warning instead.

`yarn run jest -c=test/client/jest.config.js client/state/themes/test/actions.js` to run the test directly

**Comment by sixhours** [comment]
> I'm not completely sure what's going on, but it's possibly that calling activate on a theme in a test used to fetch its data, but after the changes made, it does not, it only shows a warning instead.

I was wondering about this -- it's funny because the site ID we pass to the test is a very old site using a retired theme. I bet that could product unexpected results when we show a modal but we're expecting to activate a theme instead. I'll dig into this more.

**Comment by sixhours** [comment]
Nope, changing the site ID to one that uses a theme that is not already retired doesn't seem to make a difference. I'm stumped.

**Comment by mreishus** [comment]
is-using-retired-theme is crashing when it calls get-active-theme, because activeThemes are not defined and it assumes that activeThemes[siteId] exists

```diff
diff --git a/client/state/themes/selectors/get-active-theme.js b/client/state/themes/selectors/get-active-theme.js
index 5a7afdff91..fa05b88cac 100644
--- a/client/state/themes/selectors/get-active-theme.js
+++ b/client/state/themes/selectors/get-active-theme.js
@@ -19,7 +19,10 @@ import 'calypso/state/themes/init';
  * @returns {?string}         Theme ID
  */
 export function getActiveTheme( state, siteId ) {
+       console.log( { activeThemes: state.themes.activeThemes, siteId } );
+       console.log( 'before crash 2' );
        const activeTheme = state.themes.activeThemes[ siteId ] ?? null;
+       console.log( 'after crash 2' );
        // If the theme ID is suffixed with -wpcom, remove that string. This is because
        // we want to treat WP.com themes identically, whether or not they're installed
        // on a given Jetpack site (where the -wpcom suffix would be appended).
diff --git a/client/state/themes/selectors/is-using-retired-theme.js b/client/state/themes/selectors/is-using-retired-theme.js
index ee052390e0..05de6623b7 100644
--- a/client/state/themes/selectors/is-using-retired-theme.js
+++ b/client/state/themes/selectors/is-using-retired-theme.js
@@ -14,5 +14,9 @@ import 'calypso/state/themes/init';
  * @returns {?boolean}          Whether the current theme is retired.
  */
 export function isUsingRetiredTheme( state, siteId ) {
-       return getCanonicalTheme( state, siteId, getActiveTheme( state, siteId ) )?.retired;
+       console.log( 'before crash' );
+       console.log( getActiveTheme( state, siteId ) );
+       console.log( 'after crash' );
+       // console.log( getCanonicalTheme( state, siteId, getActiveTheme( state, siteId ) ) );
+       return false;
 }
```

**Comment by Aurorum** [comment]
Thanks for looking into this. I'm still unclear as to how other selectors can rely on `getActiveTheme` but I can't here though.

https://github.com/Automattic/wp-calypso/blob/38c37bbfe35b8c741c4e346637987ddc0fb3ab59/client/state/selectors/should-customize-homepage-with-gutenberg.js#L17

I feel like I'm probably missing something quite obvious but I'm not sure what?

**Comment by mreishus** [comment]
It may have to do with [the state set up](https://github.com/Automattic/wp-calypso/blob/d391a99110953bdf224dd230797e37ea069792f2/client/state/themes/test/actions.js#L574-L587) in the tests. I think adding safe navigation operators may be sufficient:
```diff
diff --git a/client/state/themes/selectors/get-active-theme.js b/client/state/themes/selectors/get-active-theme.js
index 5a7afdff91..d23e4d8c7b 100644
--- a/client/state/themes/selectors/get-active-theme.js
+++ b/client/state/themes/selectors/get-active-theme.js
@@ -19,7 +19,7 @@ import 'calypso/state/themes/init';
  * @returns {?string}         Theme ID
  */
 export function getActiveTheme( state, siteId ) {
-       const activeTheme = state.themes.activeThemes[ siteId ] ?? null;
+       const activeTheme = state.themes?.activeThemes?.[ siteId ] ?? null;
        // If the theme ID is suffixed with -wpcom, remove that string. This is because
        // we want to treat WP.com themes identically, whether or not they're installed
        // on a given Jetpack site (where the -wpcom suffix would be appended).

```
You'd want to test this out when actually using the site as well

**Comment by Aurorum** [comment]
Thank you! That seems to work while still ensuring everything functions the same. 

Screenshot 2021-07-23 at 14 13 48

**Comment by sixhours** [comment]
Yay, tests are passing! But I found another bug. 🙃 I don't think we can ship this until it's fixed.

When you switch from a retired theme to another non-auto-loading-homepage theme in the Trending tab, the modal doesn't appear and the theme doesn't activate. This behavior is working on production but not with this PR. Going to dig into it and see if I can find the cause.

**Comment by sixhours** [comment]
Urgh, it's related to the bug I patched temporarily in #54648 -- updating the number of themes to show by default to a ridiculous number like 300 fixes the bug.

Because Twenty Eleven appears in the All Themes view somewhere within the top 300 themes, but not the top 30, so it doesn't "count" when viewing a non-paginated static query like Trending, even if it appears close to the top of that query.

**Comment by sixhours** [comment]
I spent all afternoon chasing this bug down, and I think I finally found it.

I've been able to trace this back to `installingTheme` being `null` when we call it from a theme that isn't yet loaded into state. The modal doesn't get passed an `installingTheme` object so it never opens:

https://github.com/Automattic/wp-calypso/pull/40793/files#diff-a1a31bc09661de2bd55c1207c918ade388a6c3bcc7932a369893bc9247bc2efaR165

I confirmed the dispatch action is happening as intended and passing the correct `installingThemeId`. However, `getTheme` returns null for that ID because the query has only loaded the first 30 results into state ([value controlled by this number](https://github.com/Automattic/wp-calypso/pull/54648/files#diff-3e59937e475ddbe025a5e9ad02ba04982a7872f910c99164212098f138ac899cR223)):

https://github.com/Automattic/wp-calypso/blob/trunk/client/state/themes/selectors/get-theme.js#L23

Any theme outside the first 30 results for that query is `null` and never gets loaded into state unless you visit All Themes and scroll down to the bottom of the Showcase (which refreshes the query and updates state with the data for every single theme).

This may be partly fixed in this PR by rebasing to include [this recent bug fix](https://github.com/Automattic/wp-calypso/pull/54770) -- the query might be pulling back the wrong first 30 themes because it's a merged query with the Trending themes and Recommended themes -- but it won't fix the larger issue, which is finding a way to load all the theme results for a particular query into state. That could involve enabling pagination for custom queries, or setting the static integer value high enough to include all the themes we expect to pull back in a particular custom query.

I think in the interim, since we have a relatively small number of custom queries, we can pass a prop to `ThemesSelection` that specifies the number we expect to pull back. In the case of Trending themes, that would be all of them, for Recommended themes, it's about 30. For FSE themes, it's about 5, etc. It's hacky, but I suspect it will be faster and easier than trying to enable pagination on custom queries.

**Comment by sixhours** [comment]
This PR also breaks theme activations for FSE themes when you enable the flag `?flags=gutenboarding/site-editor`

Again, because those themes aren't pulled into state with the initial Recommended query. 🙃 I think rebasing to include #54770 might fix that, though.

**Comment by Aurorum** [comment]
Sorry for the additional ping while resolving conflicts - didn't realise that would happen! This doesn't impact that codebase, it was just for updating the branch.

**Comment by Aurorum** [comment]
Thanks for taking the time to look into this @sixhours! I've included #54770 as suggested, and I think that's fixed the issues you've mentioned, but I must admit that I never noticed them to begin with so I'm not certain.

Screenshot 2021-07-25 at 12 04 06
Screenshot 2021-07-25 at 12 04 39

**Comment by sixhours** [comment]
Unfortunately it's still breaking activations for some themes when you're switching away from a retired theme using the Trending section. 🙃 I don't know if this is edge-case enough for it to be considered a blocker.

@kwight what do you think? Continue holding on this or ship it and document the bug? I'm hoping @mreishus forays into using React Query will have the happy side effect of fixing this bug.

**Comment by kwight** [comment]
@sixhours If the broken behaviour isn't in prod, let's not introduce the regression.

If I understand correctly, the issue is that it's possible to have an empty `installingTheme` object if switching from (or to, in the case of FSE?) a theme that's not in state (whether retired, or not part of the current tab context)? (If that's true, where does the current theme info come from, a different part of the state tree?)

Hm, retired or not, it seems like the current theme needs to be where it's expected in the state tree, and FSE themes would all be in state for the tab to display properly (and a theme to get selected)? @blackjackkent @mreishus is there something we're missing here? 🙃

**Comment by kwight** [comment]
I'm so sorry @Aurorum ; after some internal discussion about retired themes and tooling, I think we're better off leaving this dialog as-is (so, dropping this PR). Again, so sorry this one didn't work out, we do really appreciate all of your work. ❤️

**Comment by Aurorum** [comment]
No worries - thanks to everyone who spent time helping to try and figure this out, and sorry it couldn't work in the end. :)

**Pull request closed without merging**
